data/impl/RecordSet.js

```diff
@@ -63,20 +63,45 @@ export class RecordSet {
         if (!filter) return this;
 
         const passes = new Map(),
-            {records} = this;
-
-        // A record that passes the filter also recursively passes all its parents.
-        const markPass = (rec) => {
-            if (passes.has(rec.id)) return;
-            passes.set(rec.id, rec);
-            const {parent} = rec;
-            if (parent) markPass(parent);
-        };
-
-        records.forEach(rec => {
-            if (filter(rec)) markPass(rec);
+            isMarked = (rec) => passes.has(rec.id),
+            mark = (rec) => passes.set(rec.id, rec);
+
+        // Pass 1.  Mark all passing records, and potentially their children recursively.
+        // Any row already marked will already have all of its children marked, so check can be skipped
+        const includeChildren = withDefault(filter.includeChildren, false);
+        let markChildren;
+        if (includeChildren) {
```

Comment:
Don't think we need to create this func conditionally since we only call it when `includeChildren` is true? Somewhat wasteful to create the func if we don't need it but the code might be slightly cleaner.

Reply:
I had the cleaner code, and mucked it up for just the reason you mentioned!  

Thinking it through, it now seems silly (and wrong) to me, given that the cost to setup the filter is going to be small relative to the cost to execute it.

Good catch